### DISCUSSION TITLE
Force `npm update` to update package.json (same as npm 6.x before)

### DISCUSSION THREAD

**Discussion by AcidSlide-Octal** [npm/feedback#270] [CLI] [answered]
Hi there,

Is there a way (with npm 7.x) to save/update the `package.json` when doing an `npm update`?

I know before in npm 6.x this was the default, and there was `--no-save` if you don't want the `package.json`.  I'm looking for the reverse of that in version 7.x.

Thanks in advance.

**Comment by dsnoeck** [comment]
I found interresting to have the `package.json` updated on `npm update` because it help to see what has been updated faster in Pull request. At least if we could have an option like --save to apply those changes, it would be great.

**Reply by AcidSlide-Octal** [reply]
exactly my point.. at least a way other devs know that the packages have been updated

**Reply by n1ngu** [reply]
For this concern, I'll usually parse the output of the npm list command, and format it for human review like
```
npm list --unicode=false --depth=9999 | tail -n +2  | sed  -e 's:^.*-- ::g' -e 's: deduped$::g' | grep -vE '^$' | sort | uniq > package-list.txt
```
It is cumbersome but it gives much more insight than the package.json and a cleaner view than the package-lock.json. You can also cut the dependency tree to the depth you are interested in. At depth 1 it becomes the resolution for your direct requirements only, but I like to watch the whole tree.

**Comment by MylesBorins** [comment]
pinging @npm/cli

**Reply by AcidSlide-Octal** [reply]
any update from @npm/cli ?

**Comment by AcidSlide-Octal** [comment]
Bumping issue/request.

**Reply by MylesBorins** [reply]
@AcidSlide-Octal I've prodded the team internally but you might want to try and open an issue over at https://github.com/npm/cli as well

**Comment by darcyclarke** [comment, marked as answer]
Today there's no way to do this for all deps easily in `v7` - `npm install` is the right tool to update `package.json` but we've definitely heard from folks that they like the ergonomics of being able to update **all** their dep versions at the same time not just in `package-lock.json` (as you've noted); I've gone ahead & added this as a topic for our [Open RFC Call today](https://github.com/npm/rfcs/issues/346) & can circle back if there is any outcomes from that discussion (ie. if we think a new proposal could be written that adds this functionality back into `update` or somehow into `install`, most likely through flags)

**Reply by cawa-93** [reply]
So ... What news? The community [**really** wants to hear the answer](https://github.com/npm/cli/issues/708).

**Reply by n1ngu** [reply]
@cawa-93 from that very issue, I'd like to highlight this comment: https://github.com/npm/cli/issues/708#issuecomment-840725231

I agree with the general idea: there is little point in automatically evolving package.json and a mismatch with the versions in package-lock.json is perfectly fine. Major version upgrades always needed a manual action anyway and minor/patch updates should only be a concern for the lockfile.

**Reply by arash-bizcover** [reply]
@n1ngu  Even if you say a little point it was a good and useful point, why removing any point at all ?

**Reply by ljharb** [reply]
@n1ngu thats not accurate. If a dep adds a new feature in v1.1 and you use it, you can’t continue to depend on 1.0 - you must update it to v1.1. Same goes with a bug fix you rely on.

Updates for packages are in no way confined to majors.

**Reply by n1ngu** [reply]
@ljharb precisely in that case, you would be _manually_ bumping the spec because you evolved your codebase to adopt a new feature, so the new version truly is a requirement of that development. A human should state that, not a robot.

If a new version shows up and it is simply about unused features, refactors, performance improvements or whatnot... running npm update and evolving the lockfile for your internal test suite is the only thing you should do. Bumping your requirement specs would be a fraud: your code works just fine with the old version and you would only complicate the dependency resolution process for your library users.

Sorry for the strong language, no offense intended :heart:

@arash-bizcover I was being soft with my words, I think there is no point at all to _automatically_ evolve the package.json dependency specs.

**Reply by piejanssens** [reply]
You assume that the dev has knowledge that this new feature was released in a later version than the one specified in package.json and therefore said dev needs to manually run `npm i <package>` to update package.json. The dev might not be aware that this is a new feature at all.

**Reply by IdeaNext** [reply]
@n1ngu I realize the logic behind this breaking change and of course a human modification in those situations is better than a robot , but isn't the caret sign (^)  in package.json versions is the way to accomplish this ? If a human stated that a package should always be in it's latest version (^), wouldn't you expect it to be this way *automatically*?

**Reply by n1ngu** [reply]
@IdeaNext I fear that is a misunderstanding of the caret version specifier. With a caret you are not requiring the latest but a **compatible version** with whatever follows the caret according to semver. That is either from the smallest patch for that minor to whatever minor is published years after for the same major.

If you are distributing a library, you should not expect control over what your dependants resolve, and they will thank you for not being picky with your dependency specs (deduplicating their resolved dependencies). If you are distributing an app you should be deploying a shrinkwrap anyway so whatever is specified in the package.json is not an issue. :woman_shrugging:

**Reply by ruyadorno** [reply]
Please note that starting with `npm@8.3.2` there's now an option to use `npm update --save` in order to save updated dependencies semver ranges back to the `package.json` file. It's also possible to have `save=true` in your `.npmrc` in order to make that the default behavior.

**Reply by matthewmayer** [reply]
i think there may be a bug in the new behavior regarding exact versions, see my comment at https://github.com/npm/cli/issues/2704#issuecomment-1018400947

**Reply by matthewmayer** [reply]
@ruyadorno could you review https://github.com/npm/cli/issues/2704#issuecomment-1018400947 and see if this is intentional?

**Comment by AcidSlide-Octal** [comment]
@darcyclarke thanks for the letting me join the Open RFC session, i'll discuss in the thread in the next few days our main Use Case scenario (which is used currently in couple of our main projects)
